### PR TITLE
Story/ccls 1994 prior authority screens

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFlowFormData.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ */
+@Data
+public class PriorAuthorityFlowFormData {
+
+  /**
+   * The details of the prior authority type form data.
+   */
+  private PriorAuthorityFormDataTypeDetails priorAuthorityTypeFormDataDetails;
+
+  private PriorAuthorityFormDataDetails priorAuthorityFormDataDetails;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFlowFormData.java:17]: javadoc

File: src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFormDataDetails.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.NoArgsConstructor;
+
+/**
+ * This class represents the details of a prior authority details form data.
+ */
+@Data
+@NoArgsConstructor
+public class PriorAuthorityFormDataDetails {
+
+  private String summary;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFormDataDetails.java:16]: javadocs

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java [hunk 3]
```diff
@@ -4,6 +4,7 @@
 import org.springframework.stereotype.Component;
 import org.springframework.validation.Errors;
 import uk.gov.laa.ccms.caab.bean.costs.CostsFormData;
+import uk.gov.laa.ccms.caab.bean.priorauthority.PriorAuthorityFormDataDetails;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java:7]: Unused?

File: src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFlowFormData.java (hunk 1 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFlowFormData.java:17]: amended

File: src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFormDataDetails.java (hunk 2 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/priorauthority/PriorAuthorityFormDataDetails.java:16]: amended

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java (hunk 3 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/costs/CostDetailsValidator.java:7]: amended

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/priorauthority/PriorAuthorityDetailsValidator.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * Validates the prior authority type details in the
+   * {@link uk.gov.laa.ccms.caab.bean.priorauthority.PriorAuthorityFormDataDetails}.
+   *
+   * @param target The object to be validated.
+   * @param errors The Errors object to store validation errors.
+   */
+  @Override
+  public void validate(final Object target, final Errors errors) {
+    final PriorAuthorityFormDataDetails priorAuthorityTypeDetails =
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/bean/validators/priorauthority/PriorAuthorityDetailsValidator.java:50]: Is this variable (and class?) named correctly? You have another validator called PriorAuthorityTypeDetailsValidator.

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java [hunk 5]
```diff
@@ -88,6 +88,7 @@ public Mono<CommonLookupDetail> getCommonValues(final String type, final String
             .queryParamIfPresent("code", Optional.ofNullable(code))
             .queryParamIfPresent("description", Optional.ofNullable(description))
             .queryParamIfPresent("sort", Optional.ofNullable(sort))
+            .queryParam("size", 1000)
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:91]: This goes towards resolving a couple of bugs in the backlog, CCLS-1740 and CCLS-2016. But I guess we need to cover the other ebs-api calls before we can close them.

File: src/main/java/uk/gov/laa/ccms/caab/controller/HomeController.java [hunk 6]
```diff
@@ -1,5 +1,6 @@
 package uk.gov.laa.ccms.caab.controller;
 
+import static uk.gov.laa.ccms.caab.constants.SessionConstants.APPLICATION_ID;
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/HomeController.java:3]: Unused

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 7]
```diff
@@ -1104,7 +1119,7 @@ public String scopeLimitationConfirmPost(
       if (scopeLimitationFlow.getScopeLimitationIndex() == null) {
         scopeLimitations.add(scopeLimitation);
       } else {
-        int index = scopeLimitationFlow.getScopeLimitationIndex();
+        final int index = scopeLimitationFlow.getScopeLimitationIndex();
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:1122]: final on a primitive?

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 8]
```diff
@@ -1271,4 +1286,292 @@ public String caseCostsPost(
     return "redirect:/application/proceedings-and-costs#costs";
   }
 
+  /**
+   * Displays the prior authority type selection page.
+   *
+   * @param model the model to add attributes to for the view.
+   * @return the view name for prior authority type selection.
+   */
+  @GetMapping("/application/prior-authorities/add/type")
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:1295]: It looks a bit wrong having the action type in the middle of the path. Should it not be at the end?

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java [hunk 9]
```diff
@@ -1271,4 +1286,292 @@ public String caseCostsPost(
...
+
+  /**
+   * Displays the prior authority details page for adding or editing.
+   *
+   * @param priorAuthorityAction the action (add or edit) being performed.
+   * @param priorAuthorityFlow the flow data for prior authority actions.
+   * @param model the model for adding attributes to the view.
+   * @return the view name for prior authority details.
+   */
+  @GetMapping("/application/prior-authorities/{action}/details")
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:1362]: It looks a bit wrong having the action type in the middle of the path. Should it not be at the end?

File: src/main/java/uk/gov/laa/ccms/caab/bean/validators/priorauthority/PriorAuthorityDetailsValidator.java (hunk 4 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/bean/validators/priorauthority/PriorAuthorityDetailsValidator.java:50]: amended

File: src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java (hunk 5 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/EbsApiClient.java:91]: to be adjusted properly in those stories

File: src/main/java/uk/gov/laa/ccms/caab/controller/HomeController.java (hunk 6 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/HomeController.java:3]: amended

File: src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java (hunk 9 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/application/summary/EditProceedingsAndCostsSectionController.java:1362]: This is the same as how proceedings and cost is done

File: src/main/resources/templates/application/prior-authority-details.html [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+      <span class="govuk-caption-l">Add/edit prior authority</span>
+      <h1 class="govuk-heading-l">
+        Prior authority - Step 2 of 2
+      </h1>
+      <p class="govuk-body-l">
+        Please complete the details below for your prior authority request.
+      </p>
+      <div class="govuk-grid-row">
+        <div class="govuk-grid-column-two-thirds">
+          <!--      TODO replace with summary list fragment once Pozza has implemented it-->
```
Comment (porritta) [inline src/main/resources/templates/application/prior-authority-details.html:25]: summary-fragments now exists

File: src/main/resources/templates/application/prior-authority-details.html [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                        th:errorclass="govuk-select--error"
+                        th:attr="onchange=|document.getElementById('dynamicOptions[${option.code}].fieldValueDisplayValue').value = this.options[this.selectedIndex].text;|">>
+                  <option value="">Please Select</option>
+                  <option th:each="item : ${__${option.code}__}"
+                          th:value="${item['code']}"
+                          th:text="${item['description']}">
+                  </option>
+                </select>
+              </th:block>
+              <th:block th:if="${option.dataType == 'FTS'}">
```
Comment (porritta) [inline src/main/resources/templates/application/prior-authority-details.html:74]: Can we extract this into a YAIM file?

;-)

File: src/main/resources/templates/application/prior-authority-remove.html [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+</head>
+<body>
+<div class="govuk-width-container">
+  <div id="breadcrumbs">
+    <a th:href="@{/application/proceedings-and-costs#prior-authority}" class="govuk-back-link">Cancel and return to proceedings and costs</a>
+  </div>
+  <div id="main-content">
+    <form action="#" th:action="@{/application/prior-authorities/{priorAuthorityId}/remove(priorAuthorityId=${priorAuthority.id})}" method="post">
+      <h1 class="govuk-heading-l">
+        Prior Authority Request
```
Comment (porritta) [inline src/main/resources/templates/application/prior-authority-remove.html:14]: Request?

File: src/main/resources/templates/application/prior-authority-type.html [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+<html lang="en"
+      th:replace="~{layout :: layout(title=~{::title},mainContent=~{::#main-content},pageCategory=${'cases'},breadcrumbs=~{::#breadcrumbs})}"
+      xmlns:th="http://www.thymeleaf.org" xmlns="http://www.w3.org/1999/html">
+<head>
+  <title>LAA CCMS - Create application</title>
+</head>
+<body>
+<div class="govuk-width-container">
+  <div id="breadcrumbs">
+    <a href="/application/proceedings-and-costs#prior-authority" class="govuk-back-link">Cancel and return to Proceedings and costs</a>
```
Comment (porritta) [inline src/main/resources/templates/application/prior-authority-type.html:11]: lowercase p

**Review by porritta — CHANGES_REQUESTED** [review]
comments added

File: src/main/resources/templates/application/prior-authority-type.html (hunk 13 above)
Comment (PhilDigitalJustice) [reply src/main/resources/templates/application/prior-authority-type.html:11]: amended

File: src/main/resources/templates/application/prior-authority-remove.html (hunk 12 above)
Comment (PhilDigitalJustice) [reply src/main/resources/templates/application/prior-authority-remove.html:14]: it was the same as the PUI, amended to 'Remove prior authority'

File: src/main/resources/templates/application/prior-authority-details.html (hunk 11 above)
Comment (PhilDigitalJustice) [reply src/main/resources/templates/application/prior-authority-details.html:74]: Anything you want me to do here? Not sure what else we would do?

Comment (porritta) [reply src/main/resources/templates/application/prior-authority-details.html:74]: Ouch. Was a ROOF joke. I will resolve this conversation immediately.

File: src/main/resources/templates/application/prior-authority-details.html (hunk 10 above)
Comment (PhilDigitalJustice) [reply src/main/resources/templates/application/prior-authority-details.html:25]: amended

**Review by porritta — APPROVED** [review]
LGTM

**Pull request merged**